### PR TITLE
Run phan over code base

### PR DESCRIPTION
**Pull request by LukasReschke**
First step to introduce https://github.com/etsy/phan as CI step. The current config is really permissive but a good start to get this included into our code base and found already multiple issues.

- [x] Requires https://github.com/nextcloud/3rdparty/pull/58
- [x] Undefined storage https://github.com/nextcloud/server/pull/5776#issuecomment-316162198
- [x] Add workaround for https://github.com/etsy/phan/issues/1033

### CODE REVIEW

**Comment by MorrisJobke** [comment]
Clickable link for the lazy people: https://github.com/etsy/phan

**Comment by LukasReschke** [comment]
@icewind1991 any comment about the following?
```
lib/private/Files/View.php:374 PhanUndeclaredClassMethod Call to method __construct from undeclared class \OC\Files\Storage\Storage\local
lib/private/Files/View.php:375 PhanUndeclaredClassMethod Call to method readdir from undeclared class \OC\Files\Storage\Storage\local
```
screen shot 2017-07-17 at 18 42 31

File: apps/sharebymail/lib/ShareByMailProvider.php [hunk 1]
```diff
@@ -923,7 +924,7 @@ protected function createShareObject($data) {
 	private function getNode($userId, $id) {
 		try {
 			$userFolder = $this->rootFolder->getUserFolder($userId);
-		} catch (NotFoundException $e) {
+		} catch (NoUserException $e) {
```
Comment (LukasReschke) [inline apps/sharebymail/lib/ShareByMailProvider.php]: @schiessle The NotFoundException doesn't exist and the function also throws a NoUserException. Care to comment? THX!

File: lib/private/Files/View.php [hunk 2]
```diff
@@ -1292,7 +1292,7 @@ private function getUserObjectForOwner($ownerId) {
 	 * @param \OC\Files\Storage\Storage $storage
 	 * @param string $internalPath
 	 * @param string $relativePath
-	 * @return array|bool
+	 * @return ICacheEntry|bool
```
Comment (LukasReschke) [inline lib/private/Files/View.php]: @icewind1991 Can you confirm? THX.

**Comment by MorrisJobke** [comment]
> @icewind1991 any comment about the following?

Looks like it is only called here: https://github.com/nextcloud/server/blob/c54a59d51eb924e77df6f6d2bb79ea466d169221/lib/private/Files/Filesystem.php#L654-L654 and this is never called.

**Comment by MorrisJobke** [comment]
Same with `opendir()` btw 🙈

File: build/.phan/config.php [hunk 3]
```diff
@@ -0,0 +1,155 @@
...
+	//
+	// Thus, both first-party and third-party code being used by
+	// your application should be included in this list.
+	'directory_list' => [
+		'build/.phan/stubs',
+		'3rdparty',
+		'lib/composer',
+		'themes',
+		'lib/',
+		'apps/files_external/3rdparty',
```
Comment (MorrisJobke) [inline build/.phan/config.php]: Why is this added and apps as well?

Comment (LukasReschke) [inline build/.phan/config.php]: Order matters here for the inclusion, the files_external app has dependencies on those and thus those needs to be parsed first.

**Comment by LukasReschke** [comment]
Unused storage functions removed with https://github.com/nextcloud/server/pull/5776/commits/fb452668b5a7856373a143bb3bbe6870136028f0

**Comment by LukasReschke** [comment]
This is ready to review now 😄

**Review by MorrisJobke — APPROVED** [review]

**Review by rullzer — APPROVED** [review]
